File: c++/medium/bear_and_steady_jene.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;
int steadyGene(string gene) {

    int n=gene.size();

    unordered_map<char,int> cnt;

    for(char c:gene)
        cnt[c]++;

    int req=n/4;

    int left=0;
    int ans=n;

    for(int right=0;right<n;right++){

        cnt[gene[right]]--;

        while(
        cnt['A']<=req &&
        cnt['C']<=req &&
        cnt['G']<=req &&
        cnt['T']<=req){

            ans=min(
            ans,
            right-left+1);

            cnt[
            gene[left]
            ]++;

            left++;
        }
    }

    return ans;
}

File: python/medium/bear_and_steady_jene.py
def steadyGene(gene):

    n=len(gene)

    cnt={}

    for c in gene:
        cnt[c]=cnt.get(c,0)+1

    req=n//4

    left=0

    ans=n

    for right in range(n):

        cnt[gene[right]]-=1

        while(
        cnt.get('A',0)<=req and
        cnt.get('C',0)<=req and
        cnt.get('G',0)<=req and
        cnt.get('T',0)<=req
        ):

            ans=min(
            ans,
            right-left+1)

            cnt[
            gene[left]
            ]+=1

            left+=1

    return ans